### PR TITLE
loops-for_loops-hard-q6

### PR DESCRIPTION
**Pull request by ketphan02**
## PR Checklist
- [ ]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [ ]  Are the assertion fail messages good?
    - [ ]  Logical (does it make sense?)
    - [ ]  Readable (can a student understand what’s wrong?)
    - [ ]  Not too feed-y (does it give away the answer?)
    - [ ]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [ ]  Specific (does the fail message relate to the question?)
- [ ]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [ ]  EOF newlines
- [ ]  Do the comments for Parsons follow the correct convention?
- [ ]  Is the question being tested sufficiently?
    - [ ]  Are there representative cases to test the program?
    - [ ]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [ ]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”
---
## Website
- [ ] Is the question text on the Gamification document up to date? 
    > Make sure to include the sample output if there is one for the question

### CODE REVIEW

File: src/loops/for_loops/hard/q6/MainTest.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  @ParameterizedTest
+  @MethodSource("inputProviderValid")
+  public void printsCorrectFactorial(int n, int factorial) throws InvalidClauseException {
+    TestOption.incorrectStructureErrorMessage = "Your program does not print out the correct factorial of the given input.";
+    runWithInput(n + "", new Clause[][] { {
+        new IntegerLiteral(n),
+        new StringLiteral("\\! \\= "),
+        new IntegerLiteral(factorial),
```
[the remark this answers is not in the document]
Comment (Hedgemon4) [reply src/loops/for_loops/hard/q6/MainTest.java:56]: Name this so you can `getItemByName` for the assert

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/loops/for_loops/hard/q6/MainTest.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    } });
```
[the remark this answers is not in the document]
Comment (ketphan02) [reply src/loops/for_loops/hard/q6/MainTest.java:57]: I think I used it wrong lol

File: src/loops/for_loops/hard/q6/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  static Stream<Integer> inputProviderInvalid() {
+    return Stream.of(-1, -5, -10);
+  }
+
+  @ParameterizedTest
+  @MethodSource("inputProviderValid")
+  public void calculatesFactorialCorrectly(int n, int factorial) throws InvalidClauseException {
+    TestOption.incorrectStructureErrorMessage = "Your program does not print the output correctly.";
```
Comment (Hedgemon4) [inline src/loops/for_loops/hard/q6/MainTest.java:52]:
```suggestion
    TestOption.incorrectStructureErrorMessage = "Your program does not calculate and print the factorial of the entered number correctly.";
```

File: src/loops/for_loops/hard/q6/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        new IntegerLiteral(n),
+        new StringLiteral("\\! \\= "),
+        new IntegerLiteral(factorial),
+    } });
+  }
+
+  @ParameterizedTest
+  @MethodSource("inputProviderInvalid")
+  public void printsErrorForInvalidInput(int n) throws InvalidClauseException {
+    TestOption.incorrectStructureErrorMessage = "Your program does not print out the error message for invalid input.";
```
Comment (Hedgemon4) [inline src/loops/for_loops/hard/q6/MainTest.java:63]:
```suggestion
    TestOption.incorrectStructureErrorMessage = "Your program does not print out an error message for invalid input.";
```

File: src/loops/for_loops/hard/q6/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    runWithInput(n + "", new Clause[][] { {
+        new IntegerLiteral(n),
+        new StringLiteral("\\! \\= "),
+        new IntegerLiteral(factorial),
+    } });
+  }
+
+  @ParameterizedTest
+  @MethodSource("inputProviderInvalid")
+  public void printsErrorForInvalidInput(int n) throws InvalidClauseException {
```
Comment (Hedgemon4) [inline src/loops/for_loops/hard/q6/MainTest.java:62]:
```suggestion
  public void printsErrorMessageForInvalidInput(int n) throws InvalidClauseException {
```

**Review by Hedgemon4 — DISMISSED** [review]

**Review by Hedgemon4 — APPROVED** [review]

**Pull request merged**